### PR TITLE
[AIRFLOW-4396] Provide a link to external Elasticsearch logs in UI.

### PR DESCRIPTION
**Pull request by kunalyogenshah**
### Jira

- [x] My PR addresses the following:
  - https://issues.apache.org/jira/browse/AIRFLOW-4396

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
This PR adds a new button to the Airflow UI next to "View log" ("View log in ES"). It provides an external URL for visualizing the log, and renders the latest try for the Task, DAG and execution being looked up. This button is controlled by a new `elasticsearch_frontend` conf param, which provides the template URL for viewing the corresponding log. Screenshot below :

UI without frontend provided:
Screen Shot 2019-05-03 at 2 37 50 PM

This is the UI when a frontend is provided:
Screen Shot 2019-05-03 at 2 39 40 PM

Documentation update is pending code review. Will update documentation after.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Unit tested the UI and checked the console for errors in the JS. Button doesn't show up if the param is not defined.

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [ ] Passes `flake8`

### CODE REVIEW

**Comment by KevinYang21** [comment]
This is nice, thank you @kunalyogenshah ! Users using Airflow with Elasticsearch backend do likely want to utilize more those aggregation features Elasticsearch frontends, e.g. Kibana, provide. 

Before looking into the code, my opinion on the UI change is that we make the layout similar to how we do the log downloading--we allow users to choose the try number of the task instances. Do you think that make sense?

**Comment by feng-tao** [comment]
@kunalyogenshah , not sure if this is the right way to show the UI. I think another approach(Lyft) is to leverage https://github.com/apache/airflow/pull/5059 and https://github.com/apache/airflow/pull/5094 which allows you to customize your task instance panel in generic way.

**Comment by feng-tao** [comment]
And we have already did that at Lyft in prod for quite some time(https://eng.lyft.com/running-apache-airflow-at-lyft-6e53bb8fccff). cc @milton0825

**Comment by kunalyogenshah** [comment]
Ah. Thanks @feng-tao . This seems to be exactly similar to what we've been trying to do. Let me look into using extra links instead.

**Comment by kunalyogenshah** [comment]
@feng-tao : A quick naive clarification question. Does this mean that the DAG operator itself must specify the link to Kibana as an extra link in its operator definition? Or is there a way to make that universal for all DAGs and tasks?

**Comment by milton0825** [comment]
@kunalyogenshah I think this is what you are looking for: https://github.com/apache/airflow/pull/5094
We can register an extra link that is global to all the operators.

**Comment by feng-tao** [comment]
+1 for @milton0825 , cc @Fokko as well

**Comment by kunalyogenshah** [comment]
Perfect. Thanks @milton0825 . I'll move my questions over to that PR for better context :)

**Comment by kunalyogenshah** [comment]
Bringing this thread back up to share some alternate thoughts. I'd like to advocate for this approach specifically for just logs look up for the following reasons : 
 - Using elasticsearch for now is quite well baked into the log handling, and Kibana is possibly the most common frontend for this. Using that as a frontend for visualizing / searching logs seems like quite a useful central feature to have as compared to a plugin. Semantically we'd use extra links plugin for features that are more specific to each individual Airflow DAG / Task / Provider.
 - This can be quite easily integrated and extended like the existing download logs / view logs, and seems on the same plane w.r.t. utility.
 - Also referring to the proposed metrics changes in this PR (https://github.com/apache/airflow/pull/5139), having a standalone endpoint for serving logs allows us to track metrics for just that endpoint better than a common extra links section. (like tracking views and failures, etc). 

Thoughts?

**Comment by milton0825** [comment]
@kunalyogenshah I am thinking about making `extra_links` aware of `try-number`. Once that is done, we can probably consolidate all the `download logs` and `view logs` and register them through extra_links. `download logs` and `view logs` will by default be there and can be disabled through `airflow.cfg`. What do you think? 

I am fine with shipping this PR now and rework later to support what I proposed.

File: airflow/config_templates/default_airflow.cfg [hunk 1]
```diff
@@ -550,6 +550,7 @@ hide_sensitive_variable_fields = True
 elasticsearch_host =
 elasticsearch_log_id_template = {{dag_id}}-{{task_id}}-{{execution_date}}-{{try_number}}
 elasticsearch_end_of_log_mark = end_of_log
+elasticsearch_frontend =
```
Comment (KevinYang21) [inline airflow/config_templates/default_airflow.cfg:556]: NIT: We can have an example comment here to make it easier to adapt.

**Review by KevinYang21 — COMMENTED** [review]
Just to confirm, this has been tested on your local environment and you've confirmed it was working right? Do you mind share the test setup( e.g. the frontend url w/o sensitive info) please?

File: airflow/config_templates/default_airflow.cfg (hunk 1 above)
Comment (kunalyogenshah) [reply airflow/config_templates/default_airflow.cfg:556]: Done!

**Comment by kunalyogenshah** [comment]
Yeah, I tested this by just changing the cfg file to include the frontend URL. The frontend URL I used is also included in the CFG file.

**Comment by kunalyogenshah** [comment]
Tailored example URL from default cfg file. Caused parsing errors for the default file. For Kibana specifically, the format I used in testing was :

`elasticsearch_frontend = <Domain>/_plugin/kibana/app/kibana#/discover?_g=()&_a=(filters:!((query:(match:(log_id:(query:'{log_id}',type:phrase))))))`

**Review by KevinYang21 — APPROVED** [review]
Baking it for 2 more days to allow other committers to review. Then I'll merge

**Comment by mik-laj** [comment]
@kunalyogenshah This features is very interesting. Can you write documentation for this feature? The description in the configuration file is very little visible. I think you should add a note to the file: `doc/howto/write-logs.rst`.

**Comment by kunalyogenshah** [comment]
Hey @mik-laj . I can definitely write documentation if needed. Will send a followup PR.

**Comment by kurtqq** [comment]
@kaxil can this PR cheery picked to 1.10.4?

**Comment by ashb** [comment]
>  I am thinking about making extra_links aware of try-number. Once that is done, we can probably consolidate all the download logs and view logs and register them through extra_links. download logs and view logs will by default be there and can be disabled through airflow.cfg. What do you think?
>
> I am fine with shipping this PR now and rework later to support what I proposed.

@milton0825 Yes please, that would be great.

File: airflow/www/views.py [hunk 2]
```diff
@@ -1265,6 +1283,7 @@ def set_duration(tid):
 
         form = DateTimeWithNumRunsForm(data={'base_date': max_date,
                                              'num_runs': num_runs})
+        external_logs = conf.get('elasticsearch', 'elasticsearch_frontend')
```
Comment (ashb) [inline airflow/www/views.py]: For future reference this should be

```suggestion
        external_logs = conf.getboolean('elasticsearch', 'elasticsearch_frontend')
```

**Comment by ashb** [comment]
@KevinYang21 When you merge PRs can you double check the commit message correctly references the jira - the release scripts I use to make sure we get everything in to the release need it. In this case it was `[AIRFLOW]` rather than `[AIRFLOW-4396]` - you can edit the commit message in the github UI when merging. Thanks

**Pull request merged**